spectec/src/backend-prose/render.ml

```diff
@@ -24,10 +24,204 @@ let gen_macro env =
 let env config inputs outputs render_latex el prose : env =
   let symbol = Symbol.env el in
   let macro = Macro.env inputs outputs in
-  let hint = Hint.env render_latex el in
-  let env = { config; prose; symbol; macro; hint; } in
+  let env = { config; prose; render_latex; symbol; macro; } in
   env
 
+(* Translation from Al exp to El exp *)
+
+let al_to_el_unop = function
+  | Al.Ast.MinusOp -> Some El.Ast.MinusOp
+  | _ -> None
+
+let al_to_el_binop = function 
+  | Al.Ast.AddOp -> Some El.Ast.AddOp
+  | Al.Ast.SubOp -> Some El.Ast.SubOp
+  | Al.Ast.MulOp -> Some El.Ast.MulOp
+  | Al.Ast.DivOp -> Some El.Ast.DivOp
+  | Al.Ast.ExpOp -> Some El.Ast.ExpOp
+  | _ -> None
+
+let al_to_el_infixop = function
+  | "->" -> Some El.Ast.Arrow
+  | s -> Some (El.Ast.Atom s)
+
+let rec al_to_el_iter iter = match iter with
+  | Al.Ast.Opt -> Some El.Ast.Opt
+  | Al.Ast.List -> Some El.Ast.List
+  | Al.Ast.List1 -> Some El.Ast.List1
+  | Al.Ast.ListN (e, id) ->
+      let ele = al_to_el_exp e in
+      let elid = Option.map (fun id -> id $ no_region) id in
+      (match ele with
+      | Some ele -> Some (El.Ast.ListN (ele, elid))
+      | _ -> None)
+
+and al_to_el_path pl =
+  let fold_path p elp = 
+    let elp' = (match p.it with
+      | Al.Ast.IdxP ei ->
+          let elei = al_to_el_exp ei in
+          (match elei with
+          | Some elei -> Some (El.Ast.IdxP (elp, elei))
+          | _ -> None)
+      | Al.Ast.SliceP (el, eh) ->
+          let elel = al_to_el_exp el in
+          let eleh = al_to_el_exp eh in
+          (match elel, eleh with
+          | Some elel, Some eleh -> Some (El.Ast.SliceP (elp, elel, eleh))
+          | _ -> None)
+      | Al.Ast.DotP kwd ->
+          let (kwd, _) = kwd in
+          let elatom = El.Ast.Atom kwd in
+          Some (El.Ast.DotP (elp, elatom)))
+    in
+    Option.map (fun elp' -> elp' $ no_region) elp'
+  in
+  List.fold_left
+    (fun elp p -> Option.bind elp (fold_path p))
+    (Some (El.Ast.RootP $ no_region)) pl
+
+and al_to_el_exp expr =
+  let exp' =
+    match expr.it with
+    | Al.Ast.NumE i -> 
+        let ei = Int64.to_int i in
+        let eli = El.Ast.NatE (El.Ast.DecOp, ei) in
+        Some eli
+    | Al.Ast.UnE (op, e) ->
+        let elop = al_to_el_unop op in 
+        let ele = al_to_el_exp e in
+        (match elop, ele with
+        | Some elop, Some ele -> Some (El.Ast.UnE (elop, ele))
+        | _ -> None)
+    | Al.Ast.BinE (op, e1, e2) ->
+        let elop = al_to_el_binop op in
+        let ele1 = al_to_el_exp e1 in
+        let ele2 = al_to_el_exp e2 in
+        (match elop, ele1, ele2 with
+        | Some elop, Some ele1, Some ele2 ->
+            Some (El.Ast.BinE (ele1, elop, ele2))
+        | _ -> None)
+    | Al.Ast.TupE el ->
+        let elel = List.filter_map al_to_el_exp el in
```

Comment:
Are you sure you want filter_map here and below? That will just silently swallow non-convertible components instead of propagating failure.

Reply:
Previously, there was `if (List.length el = List.length elel) then Some ...` following `List.filter_map`, ensuring that all exprs in `el` were successfully translated to EL. So it prevents swallowing non-convertible components.

But, using the `let*` bindings in the above comment, I was able to tidy up this logic using options. So comparison of `List.length` is no longer used, nor `List.filter_map`.